### PR TITLE
[WIP] [Core] Add support for ControlNet LoRA

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR introduces support for ControlNet LoRA.

## Why the design? 

You might wonder why not unify everything in `load_lora_weights()`. The reason why I decided not to do that is because there might be use cases where people load UNet LoRA checkpoints as well as ControlNet LoRA checkpoints. Plus ControlNet LoRA will always depend on a ControlNet that's initialized from a UNet. 

## The current design 

Load things: 

```python
from diffusers import (
    StableDiffusionXLControlNetPipeline, 
    ControlNetModel, 
    UNet2DConditionModel
)
import torch

pipe_id = "stabilityai/stable-diffusion-xl-base-1.0"
lora_id = "stabilityai/control-lora"
lora_filename = "control-LoRAs-rank128/control-lora-canny-rank128.safetensors"


unet = UNet2DConditionModel.from_pretrained(pipe_id, subfolder="unet", torch_dtype=torch.float16).to("cuda")
controlnet = ControlNetModel.from_unet(unet).to(device="cuda", dtype=torch.float16)
controlnet.load_lora_weights(lora_id, weight_name=lora_filename, controlnet_config=controlnet.config)
```

Infer:

```python
from diffusers import AutoencoderKL
from diffusers.utils import load_image, make_image_grid
from PIL import Image
import numpy as np
import cv2

prompt = "aerial view, a futuristic research complex in a bright foggy jungle, hard lighting"
negative_prompt = "low quality, bad quality, sketches"

image = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/sd_controlnet/hf-logo.png")

controlnet_conditioning_scale = 0.5  # recommended for good generalization

vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16)
pipe = StableDiffusionXLControlNetPipeline.from_pretrained(
    pipe_id,
    unet=unet,
    controlnet=controlnet,
    vae=vae,
    torch_dtype=torch.float16,
).to("cuda")

image = np.array(image)
image = cv2.Canny(image, 100, 200)
image = image[:, :, None]
image = np.concatenate([image, image, image], axis=2)
image = Image.fromarray(image)

images = pipe(
    prompt, negative_prompt=negative_prompt, image=image,
    controlnet_conditioning_scale=controlnet_conditioning_scale,
    num_images_per_prompt=4
).images

final_image = [image] + images
grid = make_image_grid(final_image, 1, 5)
grid
```

Results don't seem to be following the conditioning image at all. I will play with the hyperparameters a bit to see how the results get affected. 

first_grid

<hr>

second_grid

@patrickvonplaten could you give this a look and comment on what I might be missing out on?

### CODE REVIEW

File: src/diffusers/loaders.py [hunk 1]
```diff
@@ -1008,19 +1007,12 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
...
         r"""
         Return state dict for lora weights and the network alphas.
 
-        <Tip warning={true}>
-
-        We support loading A1111 formatted LoRA checkpoints in a limited capacity.
-
-        This function is experimental and might change in the future.
-
-        </Tip>
```
Comment (sayakpaul) [inline src/diffusers/loaders.py:1022]: I think we have graduated from this.

**Comment by patrickvonplaten** [comment]
Hmmm, I think I would have to dive into ComfyUI code to better understand how it's supposed to work. Did you get it working in ComfyUI? I think we have to start with the original code here and then work our way into `diffusers`

**Comment by sayakpaul** [comment]
The current implementation is based whatever info I managed to get from Yan (ComfyUI). I will go through ComfyUI to see how it's supposed to work but might need a second hand here.

**Comment by patrickvonplaten** [comment]
> The current implementation is based on whatever info I managed to get from Yan (ComfyUI). I will go through ComfyUI to see how it's supposed to work but might need a second hand here.

Yes I think here we have to run the original checkpoints in Comfy and then take the logic from there

**Comment by HighCWu** [comment]
When I printed the weights of stabilityai's control-lora, I noticed that it contained some `norm` words. This means that their control-lora not only applies lora to linear and conv2d, but also additionally trains norm layers. Personally, only training the conditioning embedding and lora weights better matches the name of control-lora.

**Comment by sayakpaul** [comment]
> When I printed the weights of stabilityai's control-lora, I noticed that it contained some norm words. This means that their control-lora not only applies lora to linear and conv2d, but also additionally trains norm layers. Personally, only training the conditioning embedding and lora weights better matches the name of control-lora.

Well, those weights are likely coming from the base UNet. If you go over the changes of this PR, you will notice that `norm` params are also being populated.

**Comment by yanchaoguo** [comment]
> > When I printed the weights of stabilityai's control-lora, I noticed that it contained some norm words. This means that their control-lora not only applies lora to linear and conv2d, but also additionally trains norm layers. Personally, only training the conditioning embedding and lora weights better matches the name of control-lora.
> 
> Well, those weights are likely coming from the base UNet. If you go over the changes of this PR, you will notice that `norm` params are also being populated.

hurry up

**Comment by sayakpaul** [comment]
@yanchaoguo comments like "hurry up" don't help the maintainers of the library. We are working around the clock to ship the best possible features we can for our users. We get about a thousand GitHub notifications daily as maintainers and these comments don't help at all. 

I gently ask you to be a bit more patient. 

Thanks for your understanding.

**Comment by yanchaoguo** [comment]
how to use control-lora-revision

**Comment by sayakpaul** [comment]
@patrickvonplaten here's what I have found out so far. Note that this is entirely my understanding and I am happy to stand corrected. 

## Control-LoRA workflow as per ComfyUI

For debugging purposes, I developed a [simplified Colab Notebook](https://colab.research.google.com/drive/1S-NDshYL7N4S1ugF9Y86d3SRY-Fe0Qe-?usp=sharing) out of the official Colab Notebook provided in the ComfyUI repository. 

From there one, I narrowed it down to the following LoCs that are responsible for initializing a ControlNet model and populating it with the Control-LoRA parameters (such as the ones found [here](https://huggingface.co/stabilityai/control-lora/tree/main/control-LoRAs-rank128) and [here](https://huggingface.co/stabilityai/control-lora/tree/main/control-LoRAs-rank256)). 

_(For debugging, I am using this checkpoint: https://huggingface.co/stabilityai/control-lora/blob/main/control-LoRAs-rank128/control-lora-canny-rank128.safetensors (Canny).)_ 

* https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L291 -- delegates the provided checkpoint to ControlNet, one that is compatible with the LoRA checkpoints.
* https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L260C1-L269C141 -- loads the LoRA checkpoint for the compatible layers. You would also notice that some checkpoints from the underlying UNet (UNet of SDXL base in this case; which you can find out from the Colab as well) are also used during initialization. 
* LoRA-specific computation bits: 
  * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L190
  * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L229

Now, based on this observation, I tried to do a similar thing in this PR: 

* Point of entry is this class: `ControlLoRALoaderMixin`:
https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2517
* Then the original checkpoint is prepared, leveraging a modified version `convert_ldm_unet_checkpoint()`: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2532
* Then load the converted checkpoint into the underlying ControlNet for whatever keys that are getting matched: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2537 (following the similar approach as ComfyUI). 
* Then I populate the LoRA-specific parameters. Follow from here: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2544. 

This along with the code snippet I provided in my PR description, does follow what ComfyUI is doing. But as mentioned above, I am happy to stand corrected.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten here's what I have found out so far. Note that this is entirely my understanding and I am happy to stand corrected.
> 
> ## Control-LoRA workflow as per ComfyUI
> For debugging purposes, I developed a [simplified Colab Notebook](https://colab.research.google.com/drive/1S-NDshYL7N4S1ugF9Y86d3SRY-Fe0Qe-?usp=sharing) out of the official Colab Notebook provided in the ComfyUI repository.
> 
> From there one, I narrowed it down to the following LoCs that are responsible for initializing a ControlNet model and populating it with the Control-LoRA parameters (such as the ones found [here](https://huggingface.co/stabilityai/control-lora/tree/main/control-LoRAs-rank128) and [here](https://huggingface.co/stabilityai/control-lora/tree/main/control-LoRAs-rank256)).
> 
> _(For debugging, I am using this checkpoint: https://huggingface.co/stabilityai/control-lora/blob/main/control-LoRAs-rank128/control-lora-canny-rank128.safetensors (Canny).)_
> 
> * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L291 -- delegates the provided checkpoint to ControlNet, one that is compatible with the LoRA checkpoints.
> * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L260C1-L269C141 -- loads the LoRA checkpoint for the compatible layers. You would also notice that some checkpoints from the underlying UNet (UNet of SDXL base in this case; which you can find out from the Colab as well) are also used during initialization.
> * LoRA-specific computation bits:
>   
>   * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L190
>   * https://github.com/comfyanonymous/ComfyUI/blob/1c8ae9dbb249ed5326d61d16b4e6b5807c09c0e1/comfy/controlnet.py#L229
> 
> ## What we're doing
> Now, based on this observation, I tried to do a similar thing in this PR:
> 
> * Point of entry is this class: `ControlLoRALoaderMixin`:
>   https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2517
> * Then the original checkpoint is prepared, leveraging a modified version `convert_ldm_unet_checkpoint()`: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2532
> * Then load the converted checkpoint into the underlying ControlNet for whatever keys that are getting matched: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2537
>    (following the similar approach as ComfyUI).
> * Then I populate the LoRA-specific parameters. Follow from here: https://github.com/huggingface/diffusers/blob/86f5980ce8b83c8585c98994ba489aa7c6f8fe9c/src/diffusers/loaders.py#L2544
>   .
> 
> This along with the code snippet I provided in my PR description, does follow what ComfyUI is doing. But as mentioned above, I am happy to stand corrected.

Where does a divergence appear? If you get ComfyUI working where is the point that the official implementation starts to differ from ours. E.g. where do numerical values diverge?

**Comment by sayakpaul** [comment]
> Where does a divergence appear? If you get ComfyUI working where is the point that the official implementation starts to differ from ours. E.g. where do numerical values diverge?

For now, I am only considering the final results. I couldn't identify any implementation-related differences in our implementation which is why I tried to detail as much as possible in my previous comment.

**Comment by patrickvonplaten** [comment]
> only

Can you set intermediate print statements to find out at which point the results start to diverge?

**Comment by clarencechen** [comment]
@sayakpaul This gist (https://gist.github.com/madebyollin/e6e217a77db94e2a960ab6ccbd627db9) by @madebyollin helped him debug and fix the SDXL VAE when running in FP16. It may also be helpful with comparing and debugging the Control LoRA activations.

**Comment by sayakpaul** [comment]
@clarencechen you mean we could reuse the gist to see if there's a problem FP16 inference here?

**Comment by clarencechen** [comment]
> @clarencechen you mean we could reuse the gist to see if there's a problem FP16 inference here?

I think you can reuse the gist to help track and diagnose most numerical discrepancies in Pytorch models.

**Comment by adhikjoshi** [comment]
this is not a priority anymore?

**Comment by sayakpaul** [comment]
Closing for https://github.com/huggingface/diffusers/pull/5938.

**Comment by lavinal712** [comment]
图像

I use diffusers to load control-lora canny from https://huggingface.co/stabilityai/control-lora. Here is the result.

**Comment by lavinal712** [comment]
@sayakpaul Does this meet your expected outcome?

**Pull request closed without merging**